### ISSUE TITLE
PostgreSQL arrays don't deserialise correctly when using XML MULTISET emulation

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#13872]
In PostgreSQL, when fetching an array such as `ARRAY['a', 'b']::TEXT[]`, the XML emulation may produce a query like this:

```sql
select (
  select xmlelement(
    name result,
    xmlagg(xmlelement(
      name record,
      xmlelement(name "v0", "v0")
    ))
  )
  from (
    select null as "v0"
    union all
    select cast('{"a","b"}' as varchar[]) as "a"
  ) as "t"
)
```

Which results in an XML document like this:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<result>
   <record>
      <v0 />
   </record>
   <record>
      <v0>
         <element>a</element>
         <element>b</element>
      </v0>
   </record>
</result>
```

This isn't being deserialised correctly by jOOQ. The following stack trace is produced:

```
org.jooq.exception.DataAccessException: SQL [select (select xmlelement(name result, xmlagg(xmlelement(name record, xmlelement(name "v0", "v0")))) from (select cast(? as varchar[]) as "v0" union all select cast(? as varchar[]) as "a") as "t") -- SQL rendered with a free trial version of jOOQ 3.18.0-SNAPSHOT]; Error while reading field: multiset(
  select null "a"
  union all
  select array['a', 'b'] "a"
), at JDBC index: 1
	at org.jooq_3.18.0-SNAPSHOT.POSTGRES.debug(Unknown Source)
	at org.jooq.impl.Tools.translate(Tools.java:3300)
	at org.jooq.impl.DefaultExecuteContext.sqlException(DefaultExecuteContext.java:678)
	at org.jooq.impl.CursorImpl$CursorIterator.fetchNext(CursorImpl.java:1464)
	at org.jooq.impl.CursorImpl$CursorIterator.hasNext(CursorImpl.java:1425)
	at org.jooq.impl.CursorImpl.fetchNext(CursorImpl.java:238)
	at org.jooq.impl.Tools.fetchOne(Tools.java:2592)
	at org.jooq.impl.ResultQueryTrait.fetchOne(ResultQueryTrait.java:509)
	at org.jooq.impl.Tools.attach(Tools.java:1518)
	at org.jooq.impl.DefaultDSLContext.fetchOne(DefaultDSLContext.java:4662)
	at org.jooq.impl.DefaultDSLContext.lambda$55(DefaultDSLContext.java:4682)
	at org.jooq.impl.Tools.attach(Tools.java:1518)
	at org.jooq.impl.DefaultDSLContext.fetchValue(DefaultDSLContext.java:4682)
	at org.jooq.test.all.testcases.MultisetTests.lambda$211(MultisetTests.java:1491)
	at org.jooq.test.jOOQAbstractTest.withDSLContext(jOOQAbstractTest.java:2062)
	at org.jooq.test.jOOQAbstractTest.bindAndInlineWithDSLContexts(jOOQAbstractTest.java:2055)
	at org.jooq.test.BaseTest.bindAndInlineWithDSLContexts(BaseTest.java:1728)
	at org.jooq.test.all.testcases.MultisetTests.lambda$210(MultisetTests.java:1484)
	at org.jooq.test.all.testcases.MultisetTests.lambda$453(MultisetTests.java:2984)
	at org.jooq.test.all.testcases.MultisetTests.lambda$456(MultisetTests.java:3002)
	at org.jooq.test.all.testcases.MultisetTests.testRecordLevelListenerResults(MultisetTests.java:3041)
	at org.jooq.test.all.testcases.MultisetTests.testRecordLevelListenerResults(MultisetTests.java:3029)
	at org.jooq.test.all.testcases.MultisetTests.testMultisetEmulations(MultisetTests.java:3002)
	at org.jooq.test.all.testcases.MultisetTests.testMultisetEmulations(MultisetTests.java:2984)
	at org.jooq.test.all.testcases.MultisetTests.assertMultisetDataType(MultisetTests.java:1483)
	at org.jooq.test.all.testcases.MultisetTests.assertMultisetDataType(MultisetTests.java:1461)
	at org.jooq.test.all.testcases.MultisetTests.assertMultisetDataType(MultisetTests.java:1457)
	at org.jooq.test.all.testcases.MultisetTests.testMultisetProjectionDataTypeArray(MultisetTests.java:1714)
	at org.jooq.test.jOOQAbstractTest.testMultisetProjectionDataTypeArray(jOOQAbstractTest.java:5333)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:756)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
Caused by: java.sql.SQLException: Error while reading field: multiset(
  select null "a"
  union all
  select array['a', 'b'] "a"
), at JDBC index: 1
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1590)
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.apply(CursorImpl.java:1525)
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.apply(CursorImpl.java:1)
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:144)
	at org.jooq.impl.CursorImpl$CursorIterator.fetchNext(CursorImpl.java:1449)
	... 57 more
Caused by: org.jooq.exception.DataAccessException: Could not read the XML string
	at org.jooq.impl.XMLHandler.read(XMLHandler.java:173)
	at org.jooq.impl.DefaultBinding$DefaultResultBinding.readMultisetXML(DefaultBinding.java:4109)
	at org.jooq.impl.DefaultBinding$DefaultResultBinding.lambda$2(DefaultBinding.java:4084)
	at org.jooq.impl.Tools.apply(Tools.java:2080)
	at org.jooq.impl.DefaultBinding$DefaultResultBinding.readMultiset(DefaultBinding.java:4082)
	at org.jooq.impl.DefaultBinding$DefaultResultBinding.readMultiset(DefaultBinding.java:4047)
	at org.jooq.impl.DefaultBinding$DefaultResultBinding.get0(DefaultBinding.java:4040)
	at org.jooq.impl.DefaultBinding$DefaultResultBinding.get0(DefaultBinding.java:1)
	at org.jooq.impl.DefaultBinding$InternalBinding.get(DefaultBinding.java:1025)
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1580)
	... 61 more
Caused by: java.lang.IndexOutOfBoundsException: Index 1 out of bounds for length 1
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:266)
	at java.base/java.util.Objects.checkIndex(Objects.java:359)
	at java.base/java.util.ArrayList.get(ArrayList.java:427)
	at org.jooq.impl.XMLHandler.startElement(XMLHandler.java:244)
	at oracle.xml.parser.v2.NonValidatingParser.parseElement(NonValidatingParser.java:1701)
	at oracle.xml.parser.v2.NonValidatingParser.parseRootElement(NonValidatingParser.java:461)
	at oracle.xml.parser.v2.NonValidatingParser.parseDocument(NonValidatingParser.java:407)
	at oracle.xml.parser.v2.XMLParser.parse(XMLParser.java:254)
	at oracle.xml.jaxp.JXSAXParser.parse(JXSAXParser.java:298)
	at oracle.xml.jaxp.JXSAXParser.parse(JXSAXParser.java:236)
	at org.jooq.impl.XMLHandler.read(XMLHandler.java:169)
	... 70 more
```

This seems to require a generic improvement to the XML deserialisation logic, independently of it being used by `MULTISET`. To be investigated.

**Comment by lukaseder** [comment]
Related: https://github.com/jOOQ/jOOQ/issues/13871

**Comment by lukaseder** [comment]
The task to offer exporting this format is here: https://github.com/jOOQ/jOOQ/issues/13873. But even without implementing that, we should be able to import it, because it's what PostgreSQL produces for arrays.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.18.0, 3.17.4 (https://github.com/jOOQ/jOOQ/issues/13874), and 3.16.10 (https://github.com/jOOQ/jOOQ/issues/13875)
